builtin/logical/database/secret_creds.go

```diff
@@ -0,0 +1,129 @@
+package database
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+const SecretCredsType = "creds"
+
+func secretCreds(b *databaseBackend) *framework.Secret {
+	return &framework.Secret{
+		Type:   SecretCredsType,
+		Fields: map[string]*framework.FieldSchema{},
+
+		Renew:  b.secretCredsRenew(),
+		Revoke: b.secretCredsRevoke(),
+	}
+}
+
+func (b *databaseBackend) secretCredsRenew() framework.OperationFunc {
+	return func(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		// Get the username from the internal data
+		usernameRaw, ok := req.Secret.InternalData["username"]
+		if !ok {
+			return nil, fmt.Errorf("secret is missing username internal data")
+		}
+		username, ok := usernameRaw.(string)
+
+		roleNameRaw, ok := req.Secret.InternalData["role"]
+		if !ok {
+			return nil, fmt.Errorf("could not find role with name: %s", req.Secret.InternalData["role"])
+		}
+
+		role, err := b.Role(req.Storage, roleNameRaw.(string))
+		if err != nil {
+			return nil, err
+		}
+		if role == nil {
+			return nil, fmt.Errorf("could not find role with name: %s", req.Secret.InternalData["role"])
+		}
+
+		f := framework.LeaseExtend(role.DefaultTTL, role.MaxTTL, b.System())
+		resp, err := f(req, data)
+		if err != nil {
+			return nil, err
+		}
+
+		// Grab the read lock
+		b.Lock()
+		defer b.Unlock()
+
+		// Get our connection
+		db, err := b.getOrCreateDBObj(req.Storage, role.DBName)
+		if err != nil {
+			return nil, fmt.Errorf("could not find connection with name %s, got err: %s", role.DBName, err)
+		}
+
+		// Make sure we increase the VALID UNTIL endpoint for this user.
+		if expireTime := resp.Secret.ExpirationTime(); !expireTime.IsZero() {
```

Comment:
I don't think so, but there is a branch here https://github.com/hashicorp/vault/blob/master/logical/lease.go#L44 that I wanted to guard against.